### ISSUE TITLE
Manually marking Service Bus Queue Function as Failed

### ISSUE THREAD

**Issue by garoplin** [Azure/azure-functions-python-worker#1383]
#### Question

Currently I have policy of retrying setup on function and Service Bus level. When the processing throws an error the message goes back to queue 3 times as expected (defined by property **Max delivery count**). However, for some types of exceptions I don't want to retry processing, because I know the retry won't help and it's pointless to waist computing power for that. So I can't catch the exception on the top level, check whether I should retry or no. The problem is that I still want to throw that exception after catching it, so the function execution will be marked as failed. Is there any way to mark the message as completed and then throw the exception? Or to silence the exception but add some kind of property to function execution context, so that it will be visible as failed?

**Comment by VijayStroup** [comment]
Facing this issue right now with local testing.
Tried just raising an Exception but that causes the function to fire immediately again ignoring the message lock.
One solution I found for now is marking a function as failed by doing `sys.exit(0)` which seems to have the desired outcome but this produces a weird start-up traceback but after will perform as expected. I don't feel comfortable doing this thought because I do not know what it will do when deployed.

There should be a functionality like in c# to be able to set autoComplete to False or a special exception in the azure.functions.ServiceBus class that a service bus trigger will listen to and mark it as a failed function without trying to retry the function again.

**Comment by garoplin** [comment]
Any update on this?

**Comment by JanuszNowak** [comment]
Any update on this ?

**Comment by andreyshv** [comment]
The only solution i've found for now is to move message to dead letter queue and then rethow exception. This [sample](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/samples/sync_samples/deadletter_messages_and_correct.py) can held

**Comment by alecglen** [comment]
@andreyshv I'm running into this too and think you're right that dead-lettering is the correct action. However, I'm stuck on actually implementing the call to `dead_letter_message()` as shown in that example code because it expects an `azure.servicebus.ServiceBusReceivedMessage` object instead of the `azure.functions.ServiceBusMessage` type the function provides. Any chance you have a solution to that problem handy?

**Comment by andreyshv** [comment]
@alecglen sorry, but i work with .net not python. Below is my code in c#. It's differs from python of course but idea the same.

```
internal class CheckStatusFunction
    {
        private readonly IRunner _runner;

        public CheckStatusFunction(IRunner runner)
        {
            _runner = runner;
        }

        [Function(nameof(CheckStatusFunction))]
        public async Task RunAsync(
            [ServiceBusTrigger("%ServiceBusSender:QueueName%", AutoCompleteMessages = false)] ServiceBusReceivedMessage receivedMessage,
            ServiceBusMessageActions messageActions,
            CancellationToken cancellationToken)
        {
            var message = receivedMessage.Body.ToObjectFromJson<CheckOperationMessage>();

            var isSuccess = false;
            Exception exception = null;

            try
            {
                isSuccess = await _runner.RunAsync(message, cancellationToken);
            }
            catch (Exception ex)
            {
                exception = ex;
            }

            if (isSuccess)
            {
                await messageActions.CompleteMessageAsync(receivedMessage);
            }
            else
            {
                await messageActions.DeadLetterMessageAsync(receivedMessage);
                if (exception != null)
                    throw exception;
            }
        }
    }
}
```

**Comment by alecglen** [comment]
One would hope, but unfortunately the Python worker is [not designed to support that flow currently](https://github.com/Azure/azure-functions-python-worker/issues/1172).

Ultimately for our Python Function, we worked around this by flipping the script. The Function is configured without retries so errors are dead-lettered by default. Then for the specific cases that we do see value in retrying, we catch the errors and manually add a new copy of the message to queue before exiting.

Definitely some drawbacks and wouldn't work for every case, but it's working for us.

**Comment by trin94** [comment]
Regarding the [code snippet](https://github.com/Azure/azure-functions-python-worker/issues/1383#issuecomment-2474194612) from @andreyshv, I think `ServiceBusReceivedMessage` now is supported since [this PR](https://github.com/Azure/azure-functions-python-extensions/pull/107).

It's also on PyPi: https://pypi.org/project/azurefunctions-extensions-bindings-servicebus/

The only think left - at least from that sample - seems to be `ServiceBusMessageActions`. Really looking forward to these bindings as they will make the Service Bus integration much more mature and useful 👍

**Comment by hallvictoria** [comment]
Yes, binding to a `ServiceBusReceivedMessage` is now supported. You can learn more here: [Azure Service Bus for Python](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus?pivots=programming-language-python&tabs=isolated-process%2Cextensionv5%2Cextensionv3#sdk-binding-types)

Message settlement (and the corresponding `ServiceBusMessageActions` object) is still pending, but this work is in progress. Tentative ETA is August - September.
